src/qr/dclogin_scheme.rs

```diff
@@ -0,0 +1,402 @@
+use std::collections::HashMap;
+
+use crate::config::Config;
+use crate::context::Context;
+use crate::provider::Socket;
+use crate::{contact, login_param::CertificateChecks};
+use anyhow::{bail, Context as _, Result};
+use num_traits::cast::ToPrimitive;
+
+use super::{Qr, DCLOGIN_SCHEME};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LoginOptions {
+    UnsuportedVersion(u32),
+    V1 {
+        mail_pw: String,
+        imap_host: Option<String>,
+        imap_port: Option<u16>,
+        imap_username: Option<String>,
+        imap_password: Option<String>,
+        imap_security: Option<Socket>,
+        imap_certificate_checks: Option<CertificateChecks>,
+        smtp_host: Option<String>,
+        smtp_port: Option<u16>,
+        smtp_username: Option<String>,
+        smtp_password: Option<String>,
+        smtp_security: Option<Socket>,
+        smtp_certificate_checks: Option<CertificateChecks>,
+    },
+}
+
+/// scheme: `dclogin://user@host/?p=password&v=1[&options]`
+/// read more about the scheme at https://github.com/deltachat/interface/blob/master/uri-schemes.md#DCLOGIN
+pub(super) fn decode_login(qr: &str) -> Result<Qr> {
+    let url = url::Url::parse(qr).with_context(|| format!("Malformed url: {:?}", qr))?;
+
+    let mut payload = qr
+        .get(DCLOGIN_SCHEME.len()..)
+        .context("invalid DCLOGIN payload E1")?;
+
+    // if first 2 chars are `//` remove them
+    if payload.get(0..2) == Some("//") {
+        payload = payload.get(2..).context("invalid DCLOGIN payload E2")?;
+        // todo: is there a more idiomatic way?
+    }
+
+    let addr = payload
+        .get(
+            ..payload
+                .chars()
+                .position(|c| c == '?' || c == '/')
+                .context("invalid DCLOGIN payload E3a")?,
+        )
+        .context("invalid DCLOGIN payload E3b")?;
+
+    let mut scheme = url.scheme().to_owned();
+    scheme.make_ascii_lowercase();
+
+    if scheme == "dclogin" {
```

Comment:
wanted to use the url lib to be more sure the url was correctly parsed, the payload should already start with that otherwise this function is not even called. so probably no

Reply:
Why check the scheme at all, then? But anyway, we can also leave it as is.

BTW, there is also `eq_ignore_ascii_case()`.

Reply:
to make sure the uri parsing lib worked, but that step might not be need. though using `eq_ignore_ascii_case()` could shorten code a bit so feel free to do it.

Reply:
e2b94fab